proposals/0000-template.md

```diff
@@ -0,0 +1,238 @@
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+There are mainly three aspects of this proposal: changes to the protocol, changes to the NEAR -> Eth bridge, and changes to
+staking pool contracts.
+
+## Changes to the protocol
+
+There are mainly three changes on the protocol level:
+- Every approval will contain a message signed by the secp validator key that can be used for the bridge and such signatures
+will be stored in the block body.
+- We need to introduce a new staking action that takes both an ed25519 public key and a secp256k1 public key.
+- Existing staking pool contracts need to be upgraded on the protocol level since they are permissionless and do not have
+the ability to self-upgrade.
+
+## Changes to the bridge
+
+Instead of optimistically trust the validator signatures, the smart contract on Ethereuem side, which runs a
+light client of NEAR, will fully verify the light client blocks submitted by the bridge relayers.
+
+## Changes to the staking pool contract
+
+A new staking pool contract needs to be written to account for the new staking action and also gives the owner
+the ability to upgrade to the new pool.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+In this section, we will explain further the technical details of the aforementioned changes and the upgrade path.
+
+## Changes to the protocol
+
+`Approval` will be changed to
+```rust
+pub struct Approval {
+    pub inner: ApprovalInner,
+    pub target_height: BlockHeight,
+    pub signature: Signature,
+    pub bridge_signature: Signature,
+    pub account_id: AccountId,
+}
+```
+
+and `Block` will be changed to
+
+```rust
+pub struct BlockV3 {
+    pub header: BlockHeader,
+    pub chunks: Vec<ShardChunkHeader>,
+    pub challenges: Challenges,
+    pub bridge_signatures: Vec<Option<Signature>>,
+
+    pub vrf_value: near_crypto::vrf::Value,
+    pub vrf_proof: near_crypto::vrf::Proof,
+}
+```
+
+A new staking action `StakeAction2` will be introduced:
+```rust
+pub struct StakeAction2 {
+    pub stake: Balance,
+    pub secp_public_key: PublicKey,
+    pub ed25519_public_key: PublicKey,
+}
+```
+
+The bridge signature here is a signature the following message:
+```rust
+(last_final_block_height, last_final_blocks_merkle_root, next_secp_merkle_root, next_total_stake)
+```
+
+For staking pools, we will change the contract code when it is touched. More specifically, `action_stake` will
+be modified in the following way:
+```rust
+pub(crate) fn action_stake(
+    account: &mut Account,
+    result: &mut ActionResult,
+    account_id: &AccountId,
+    stake: &StakeAction,
+    last_block_hash: &CryptoHash,
+    epoch_info_provider: &dyn EpochInfoProvider,
+) -> Result<(), RuntimeError> {
+    if account.code_hash == STAKING_POOL_V1_HASH {
+        action_deploy_contract(new_staking_pool_contract);
+    }
+    // Then proceed to do what the function does now.
+}
+```
```

Comment:
Why not replace the code by hash? Basically if we're trying to execute a code for a hash `123`, then we actually need to execute hardcoded other code.

Reply:
I am not sure what you mean here. `action_deploy_contract(new_staking_pool_contract)` does replace the contract. Do you mean we shouldn't hardcode the contract binary? Then you still need to get the binary and store it in state.